File: sshappy/service.py
from __future__ import annotations

import asyncio
import logging
import time

from .config import load_config
from .db import Database
from .server import SSServer
from .state import RuntimeState

logger = logging.getLogger(__name__)


class Service:
    def __init__(self) -> None:
        self.config = load_config()
        logging.basicConfig(
            level=getattr(logging, self.config.log_level, logging.INFO),
            format="%(asctime)s %(levelname)s %(name)s: %(message)s",
        )
        if self.config.node_id <= 0:
            raise RuntimeError("node_id/NODE_ID must be set")
        self.db = Database(self.config)
        self.state = RuntimeState()
        self.server = SSServer(
            listen_host=self.config.listen_host,
            connect_timeout=self.config.tcp_connect_timeout,
            idle_timeout=self.config.tcp_idle_timeout,
            udp_idle_timeout=self.config.udp_idle_timeout,
            state=self.state,
        )
        self.node = None
        self.last_traffic_report = 0.0
        self.last_alive_report = 0.0
        self.last_node_report = 0.0

    async def run(self) -> None:
        await self.sync_once()
        server_task = asyncio.create_task(self.server.serve_forever())
        sync_task = asyncio.create_task(self.loop())
        await asyncio.gather(server_task, sync_task)

    async def loop(self) -> None:
        while True:
            await asyncio.sleep(self.config.sync_interval_seconds)
            try:
                await self.sync_once()
            except Exception:
                logger.exception("sync cycle failed")

    async def sync_once(self) -> None:
        node = await asyncio.to_thread(self.db.load_node)
        users = await asyncio.to_thread(self.db.load_users, node)
        await self.server.apply(node, users)
        self.node = node
        now = time.monotonic()

        if now - self.last_traffic_report >= self.config.traffic_report_seconds:
            traffic = await self.state.snapshot_traffic()
            await asyncio.to_thread(self.db.report_traffic, node, traffic)
            self.last_traffic_report = now
            logger.info("traffic reported: users=%s", len(traffic))

        if now - self.last_node_report >= self.config.node_report_seconds:
            online = await self.state.online_user_count()
            await asyncio.to_thread(self.db.report_node_status, node, online)
            self.last_node_report = now
            logger.info("node status reported: online=%s", online)

        if now - self.last_alive_report >= self.config.alive_ip_report_seconds:
            alive = await self.state.snapshot_alive_ips()
            await asyncio.to_thread(self.db.report_alive_ips, node, alive)
            self.last_alive_report = now
            logger.info("alive ips reported: users=%s records=%s", len(alive), sum(len(v) for v in alive.values()))

File: internal/sstest/service.go
package sstest

import (
	"context"
	"fmt"
	"log"
	"time"
)

var processStartedAt = time.Now()

func Run(ctx context.Context, config Config) error {
	if config.NodeID <= 0 {
		return fmt.Errorf("node_id/NODE_ID must be set")
	}
	db, err := OpenDatabase(config)
	if err != nil {
		return err
	}
	defer db.Close()

	state := NewRuntimeState()
	server := NewServer(config, state)
	node, err := syncOnce(db, server, state, config, true)
	if err != nil {
		return err
	}
	go func() {
		if err := server.Serve(ctx); err != nil && ctx.Err() == nil {
			log.Printf("server stopped: %v", err)
		}
	}()

	syncTicker := time.NewTicker(time.Duration(config.SyncIntervalSeconds) * time.Second)
	defer syncTicker.Stop()
	trafficTicker := time.NewTicker(time.Duration(config.TrafficReportSeconds) * time.Second)
	defer trafficTicker.Stop()
	nodeTicker := time.NewTicker(time.Duration(config.NodeReportSeconds) * time.Second)
	defer nodeTicker.Stop()
	aliveTicker := time.NewTicker(time.Duration(config.AliveIPReportSeconds) * time.Second)
	defer aliveTicker.Stop()
	onlineWindow := onlineCountWindow(config)

	for {
		select {
		case <-ctx.Done():
			return nil
		case <-syncTicker.C:
			loaded, err := db.LoadNode()
			if err != nil {
				log.Printf("sync load node failed: %v", err)
				continue
			}
			users, err := db.LoadUsers(loaded)
			if err != nil {
				log.Printf("sync load users failed: %v", err)
				continue
			}
			if err := server.Apply(loaded, users); err != nil {
				log.Printf("apply runtime failed: %v", err)
				continue
			}
			node = loaded
		case <-trafficTicker.C:
			traffic := state.SnapshotTraffic()
			if err := db.ReportTraffic(node, traffic); err != nil {
				log.Printf("traffic report failed: %v", err)
			} else {
				log.Printf("traffic reported: users=%d", len(traffic))
			}
		case <-nodeTicker.C:
			online := state.OnlineUserCount(onlineWindow)
			if err := db.ReportNodeStatus(node, online); err != nil {
				log.Printf("node status report failed: %v", err)
			} else {
				log.Printf("node status reported: online=%d", online)
			}
		case <-aliveTicker.C:
			alive := state.SnapshotAliveIPs()
			records := 0
			for _, ips := range alive {
				records += len(ips)
			}
			if err := db.ReportAliveIPs(node, alive); err != nil {
				log.Printf("alive ips report failed: %v", err)
			} else {
				log.Printf("alive ips reported: users=%d records=%d", len(alive), records)
			}
		}
	}
}

func syncOnce(db *Database, server *Server, state *RuntimeState, config Config, report bool) (NodeInfo, error) {
	node, err := db.LoadNode()
	if err != nil {
		return node, err
	}
	users, err := db.LoadUsers(node)
	if err != nil {
		return node, err
	}
	if err := server.Apply(node, users); err != nil {
		return node, err
	}
	if report {
		traffic := state.SnapshotTraffic()
		if err := db.ReportTraffic(node, traffic); err != nil {
			return node, err
		}
		log.Printf("traffic reported: users=%d", len(traffic))
		online := state.OnlineUserCount(onlineCountWindow(config))
		if err := db.ReportNodeStatus(node, online); err != nil {
			return node, err
		}
		log.Printf("node status reported: online=%d", online)
		alive := state.SnapshotAliveIPs()
		if err := db.ReportAliveIPs(node, alive); err != nil {
			return node, err
		}
		log.Printf("alive ips reported: users=%d records=0", len(alive))
	}
	return node, nil
}

func onlineCountWindow(config Config) time.Duration {
	seconds := config.NodeReportSeconds
	if config.AliveIPReportSeconds > seconds {
		seconds = config.AliveIPReportSeconds
	}
	if config.SyncIntervalSeconds > seconds {
		seconds = config.SyncIntervalSeconds
	}
	return time.Duration(seconds*2+30) * time.Second
}
